pricegraph/src/orderbook/scalar.rs

```diff
@@ -169,3 +170,52 @@ fn assert_strictly_positive_and_finite(value: f64) -> f64 {
     debug_assert!(num::is_strictly_positive_and_finite(value));
     value
 }
+
+/// The exchange rate weight used by the pathfinding algorithm.
+#[derive(Clone, Copy, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Weight(i128);
+
+impl Weight {
+    /// Creates a new graph weight from a floating point number.
+    pub fn new(value: f64) -> Self {
+        // NOTE: We need to reserve some bits in order for the addition not to
+        // overflow, specifically we need to reserve
+        // - 16 bits for the maximum number of nodes (maximum number of tokens)
+        // - 8 bits for the [-128, 128) range
+        // - 1 bit for infinity
+        // Thats a total of 25 bits, which leaves us with 103 bits of precision
+        // for the weight.
+        const FACTOR: f64 = (1u128 << 103) as _;
+
+        let weight = value.log2() * FACTOR;
```

Comment:
At first I thought this would introduce a lot error because of the multiplication with a large number. This isn't the case because the number is a power of 2 so in the float representation it only changes the exponent. This should be commented.

Since f64 uses 53 bits for the fractional part we don't get any benefit by shifting up to 103. Let's document this too and use a lower shift amount.